### PR TITLE
Timed order batching with retry behaviour

### PR DESCRIPTION
**Pull request by ScaredIbis**
# Motivation
- The order batch queueing system should handle matches coming in while in the middle of processing and retry if calls to the smart contract fail

# Changes
- Add `OrderBatcher` which uses [bottleneck](https://www.npmjs.com/package/bottleneck) to maintain a batch of matched orders and process them via a seperate queue with retry behaviour

### CODE REVIEW

**Review by raymogg — APPROVED** [review]

**Comment by raymogg** [comment]
This is 100x better implementation of the executioner. Bottleneck is cool too, will do some reading on it

File: OrderBatcher.js [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+            takerOrders.push(match[1])
+          }
+
+
+          await this.submitOrders(makerOrders, takerOrders, contract, gasLimit, sender)
+          this.removePendingMatches(nextBatch)
+        } catch (error) {
+          console.log(`ERROR SUBMITTING BATCH (ATTEMPT ${nextBatch.attempts})`, error.message, JSON.stringify(nextBatch, null, 2))
+          // if there are still attempts left
+          if(nextBatch.attempts < maxAttempts) {
```
Comment (ScaredIbis) [inline OrderBatcher.js:50]: should be <=

File: OrderBatcher.js [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+    return {
+      makeOrders: matches.map(match => match[0]),
+      takeOrders: matches.map(match => match[1]),
+      count: matches.length * 2
+    }
+  }
+
+  addMatch (pair) {
+    this.batcher.add(pair)
```
Comment (ScaredIbis) [inline OrderBatcher.js:99]: should push to pendingQueue first in case as batcher handler removes items from pendingQueue

File: index.js [hunk 3]
```diff
@@ -221,11 +179,17 @@ app.listen(3000, async () => {
...
     `)
+
     //Setup signing account
     const account = web3.eth.accounts.privateKeyToAccount(process.env.PRIVATE_KEY);
     web3.eth.accounts.wallet.add(account);
     web3.eth.defaultAccount = account.address;
     traderContract = new web3.eth.Contract(traderABI, process.env.TRADER_CONTRACT)
+
+    // begin processing matched orders
+    const maxRetries = process.env.MATCH_SUBMISSION_RETRIES || 2
```
Comment (ScaredIbis) [inline index.js:190]: should be cast to number

**Pull request merged**